### DISCUSSION TITLE
QUESTION: What are the options for flashing matter.js into devices

### DISCUSSION THREAD

**Discussion by mahadev-0007** [matter-js/matter.js#1198] [Q&A]
I have developed my custom controller with node.js also added a control bridge to it 
and used express js for rest api for communication and it works perfectly now.

My question is how to deploy these because I need an os based board and what are available option

Also concern for security because any one who gets access to the board can access the os and manipulate the code also

what are the available option please try to answer even if it is a silly question ;)

**Comment by Apollon77** [comment]
Hey, nice to hear about your project. Do you have more details or a link?

In general "any platform that supports node.js" works for matter.js ... also other patforms with JS could work but you might need to implement platform specific code.

Also your raised security concern is a topic then, so you might want to work with encrpted filesystem (at least partially) and such and you also might need an update mechanism.

But in fact beside these general infos I do not have any concrete "secured production ready" options other then "a raspi based board" as idea where you need to search yourself.

I also converted that as a discussion.

**Comment by mahadev-0007** [comment]
Why was this project started
Was this for matter apps in web and mobile or be built in products

Because only this repo is good for anything matter related
the python, rs, go, or even c++ is not even readable or dont have any documentation to start as a beginer

I was able completely understand matter and even create a controller myself with matter js very fast and easy (even though my preffered language is python)

https://github.com/mahadev-0007/matterjs-hub-proto

This is my implementation of matter.js but really looking for a solution to be built on a mcu and release my own product as a hub like aqara m3 or google home etc

**Reply by lauckhart** [reply]
For an MCU you'd probably be looking at something designed specifically to host javascript like Moddable or Espressuino. AFAIK you'd be breaking new ground... You'd need to provide some platform abstractions for things like networking and crypto, and deal with any ideosycracies of the JS implementation.

A more straightforward path is the MPU route...  Any Linux MPU with modern node.js available and enough memory can run matter.js.  This is the path I'd consider, unless maybe you have specific constraints that make this infeasible?

Device security then is standard embedded linux stuff, protected firmware, chain of trust, etc.  matter.js handles application-level security for you but is entirely JS-based so doesn't play at the system level.

**Reply by Apollon77** [reply]
We would be happy to take any such "platform specific" library for special platforms, like @lauckhart wrote, also into our packages and we are there to support with any questions when you try to start such a platform adoption. We also support mobile developers to get it working with react-native and such. From experience the most relevant challenges are crypo ... test rest is mostly not too hard to map.

To the "why the project was started" the answer is: because we wanted to implement Matter without native dependencies that could also run in browser or in general anywhere where JS has a home.
